lab-judy/app/entry.js

```diff
@@ -0,0 +1,60 @@
+'use strict';
+
+//require webpack assets
+require('./scss/base.scss');
+
+//npm modules
+const cowsay = require('cowsay-browser');
+const angular = require('angular');
+
+//angular module
+//the 2 args means this is a setter app that creates actual demoApp module
+const demoApp = angular.module('demoApp', []);
+
+//angular constructs
+demoApp.controller('CowsayController', [ '$log', CowsayController]);
+
+function CowsayController($log){
+  $log.debug('init CowsayController');
+
+  this.title = 'Moooo';
+  this.history = [];
+
+  cowsay.list( (err, cowfiles) => {
+    this.cowfiles = cowfiles;
+    this.currentCow = this.cowfiles[0];
+  });
+
+  //renders cow to page with text from input field
+  this.updateCow = function(input){
+    $log.debug('cowsayCtrl.updateCow()');
+    return '\n' + cowsay.say({text: input || 'type something dammit', f: this.currentCow});
+  };
+
+  this.speak = function(input){
+    $log.debug('cowsayCtrl.updateCow()');
+    this.spoken = this.updateCow(input);
+    this.history.push(this.spoken);
+  };
+  this.undo = function(){
+    $log.debug('this.undo()');
+    this.history.pop(); //get rid of the current thing
+    this.spoken = this.history.pop() || ''; //set the last thing
```

Comment:
One quick point -- the pop method returns the item that was popped off of the array. That's why it takes 2 clicks to make undo seem like it's working -- it assigns the last spoken text to this.spoken and doesn't effectively change even though the history array has been altered. If you want to see the true last spoken text, you should pop the item off the history and then reference the final value in the array like so:
```
this.history.pop;
this.spoken = this.history[this.history.length - 1] || '';
```